### ISSUE TITLE
Certificate Error On Windows

### ISSUE THREAD

**Issue by mike-thompson-day8** [technomancy/leiningen#1800]
The fix for #1733 broke `lien.bat` on Windows  (#126 seems related)

Detailed reports: https://groups.google.com/forum/#!topic/clojure/NTuFH0YGDeo 

We've investigated a bit and found that **neither** `wget` or `curl` work on Windows without the insecure command-line params. 

As an experiment, we removed access to BOTH `wget` and `curl` from PATH and found that `powershell` works beautifully. 

So the choice is:
   1.  Put the insecure command-line parameters back in  (doesn't seem like an option)
   2.  Remove use of `wget` and `curl`, and insist that `powershell` is installed.

**Comment by bruceadams** [comment]
Requiring Powershell seems reasonable. Hasn't Powershell been included with every Windows release for the last eight or ten years now?

**Comment by mike-thompson-day8** [comment]
Powershell is been built into Windows from Win7 onwards (released July 22, 2009).  It also seems to have been backported to WinXP via service Pack 3.

Source: "Versions" in http://en.wikipedia.org/wiki/Windows_PowerShell

So, to be without powershell,  a developer would (1) need to be using XP -- ie be using a PC that was over 6 years old plus (2) they must have chosen to not install service pack 3. 

Using `powershell` the first choice downloader seems completely reasonable and seems the right response. That has to handle 99.9% of cases.  

The only question then is:  do you want to continue to use the insecure version of `wget` and `curl` as a fall back for anyone on XP?

**Comment by technomancy** [comment]
> The only question then is:  do you want to continue to use the insecure version of `wget` and `curl` as a fall back for anyone on XP?  

No; if we include any fallback it should be to secure invocations. If anyone has trouble they can fix their CA store or download in a browser manually.

**Comment by kumarshantanu** [comment]
Powershell in `lein.bat` has worked well (commenting out _wget_ and _curl_ invocations) for me. It can be installed on Windows XP/Vista and comes pre-installed on Windows 7 and higher.

**Comment by hyPiRion** [comment]
If it's easier to install PowerShell (which I suspect?) than wget/curl and that it also handles CA stores better, then I'd recommend to download via PowerShell by default and fallback on curl, then wget. It seems like wget is the one with the most issues, and I've heard of people that fixed this by just using curl instead.

**Comment by djpowell** [comment]
The version of curl that is installed with lein-win-installer does include the CA certs and ought to work fine with SSL.  If powershell always has the correct CA certs (if it is installed), then it sounds a good idea to put that as first choice, followed by curl (which should be configured correctly with lein-win-installer), followed by wget.

I'm just going to have a look at lein-win-installer tho - it hasn't changed, but it looks like it might not be working with the latest leiningen.

**Comment by pbalduino** [comment]
My two cents:

For any reason, Powershell is not working here (Win XP SP3 and Win 7), so the fallback to wget/curl is saving my skin. Maybe `lein.bat` could show a warning message with pause to annoy the user instead to fail.

I agree with @hyPiRion and @technomancy.

**Comment by Gregg8** [comment]
That's rather curious and potentially important if it fails under Windows 7 and we're thinking of making PowerShell the default.

What error message do you get?

PowerShell uses the .NET Framework and lein.bat's use of it is to call .NET's `System.Net.WebClient.DownloadFile` function, which I note has been available from the beginning.

But it would still interesting to see what version of PowerShell and .NET you have on those machines.

If you get a moment, can you use the following commands to get the versions:

```
powershell $PSVersionTable.PSVersion.toString()
dir /b %windir%\Microsoft.NET\Framework\v*
```

Note that the second command will list ALL .NET versions you have installed.

Thanks.

**Comment by winks** [comment]
My windows 7:
```
2.0

v1.0.3705
v1.1.4322
v2.0.50727
v3.0
v3.5
v4.0.30319
```

Someone on Win8.1
```
4
~some error 
```
but he said there are 2 folders in `c:\windows\Microsoft.NET`: `Framework` and `Framework64`

**Comment by pbalduino** [comment]
I was using 2.5.0

> powershell $PSVersionTable.PSVersion.toString()
3.0

> dir /b %windir%\Microsoft.NET\Framework\v*
v1.0.3705
v1.1.4322
v2.0.50727
v3.0
v3.5
v4.0.30319

> lein -v
Leiningen 2.5.0 on Java 1.7.0_51 Java HotSpot(TM) 64-Bit Server VM

> lein upgrade
The script at c:\java\lein\lein.bat will be upgraded to the latest stable
version.
Do you want to continue (Y/N)?y
Downloading latest Leiningen batch script...
'wget' is not recognized as an internal or external command, operable
program or batch file.
Failed to download
https://github.com/technomancy/leiningen/raw/stable/bin/lein.bat

Then I downloaded lein.bat 2.5.1 manually:

> lein -v
<user dir>\.lein\self-installs\leiningen-2.5.1-standalone.jar can not be
found.
You can try running "lein self-install"
or change LEIN_JAR environment variable
or edit lein.bat to set appropriate LEIN_JAR path.

> lein self-install
Downloading Leiningen now...
  % Total    % Received % Xferd  Average Speed   Time    Time     Time
 Current
                                 Dload  Upload   Total   Spent    Left
 Speed
100   408    0   408    0     0    523      0 --:--:-- --:--:-- --:--:--
523
 25 14.5M   25 3824k    0     0   289k      0  0:00:51  0:00:13  0:00:38
 389k
100 14.5M  100 14.5M    0     0   321k      0  0:00:46  0:00:46 --:--:--
 395k
        1 file moved.

> lein -v
Leiningen 2.5.1 on Java 1.7.0_51 Java HotSpot(TM) 64-Bit Server VM


2.5.1 worked like a charm in this machine. It was installed from scratch
this Tuesday.

Regards

**Comment by Gregg8** [comment]
Thanks @winks. 

So the interesting number for all those .NET versions is that you're both on a recent version (v4.0.30319)

So, did you try updating to 2.5.1 under these environments (using PowerShell)? If so, did that succeed?

Like your colleague, I have Windows 8.1 64-bit installed and that successfully downloaded the file using PowerShell (and the `dir` command worked in my case):

```
4.0

v1.0.3705
v1.1.4322
v2.0.50727
v3.0
v3.5
v4.0.30319

```

**Comment by Gregg8** [comment]
Thanks @pbalduino,

Like us others, you're on the same .NET Framework, although you're on PowerShell v3.

Thanks for the detailed output. On reviewing, it was `wget` that was failing rather than `PowerShell`, which is reassuring  for the idea of defaulting to PowerShell.

Not sure if this was your original experience or a new run though.

In any case, I see that it's now on the 2.5.2 milestone. Thanks @hyPiRion.

**Comment by winks** [comment]
So... should the default be powershell in 2.5.2 or not? Seems like the change would be just the order of trying the tools.

Apparently:

  * 2.5.1+ works if you don't have wget/curl in your path
  * you can still have wget/curl versions that don't work and shadow powershell.

**Comment by mike-thompson-day8** [comment]
Powershell works. It should definitely be the first option tried.

Then, by all means fall back to `curl` and `wget` but in my testing they don't work (unless made insecure, which is not an option).

**Comment by hyPiRion** [comment]
Happy to take a patch that defaults to powershell, 

If none of the options work, it'd be nice to give a helpful error message back which could help them fetching lein.

**Comment by fgiasson** [comment]
@hyPiRion here it is. Also note there is already a message if none of the options are available (if this is what you were referring to).

**Comment by hyPiRion** [comment]
I was thinking about a more descriptive error message – currently it only emits "Failed to download ...". Explaining that it *might* be related to certificate errors and some (safe) workarounds for that problem would be nice.

**Comment by fgiasson** [comment]
@hyPiRion ok, check my second commit: #1831

**Comment by hyPiRion** [comment]
This was fixed by #1831.
